### PR TITLE
394 fix structdict copy

### PR DESCRIPTION
**Pull request by dfsp-spirit**
Author Guidelines
-----------------
- [x] Is the change set **< 600 lines**?
- [x] Was the code checked for memory leaks/performance bottlenecks?
- [x] Is the code running locally **and** on the ESI cluster?
- [x] Is the code running on all supported platforms?

Reviewer Checklist
------------------
- [ ] Are testing routines present?
- [ ] Do **parallel** loops have a set length and correct termination conditions?
- [ ] Do objects in the global package namespace perform proper parsing of their input? 
- [ ] Do code-blocks provide novel functionality, i.e., no re-factoring using builtin/external packages possible?
- [ ] Code layout
  - [ ] Is the code PEP8 compliant?
  - [ ] Does the code adhere to agreed-upon naming conventions?
  - [ ] Are keywords clearly named and easy to understand?
  - [ ] No commented-out code?
- [ ] Are all docstrings complete and accurate?
- [ ] Is the CHANGELOG.md up to date?

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,15 @@
...
 All notable changes to this project will be documented in this file.
 
+## Unreleased WIP
+
+### NEW
+
+### CHANGED
+
+### Fixed
+- fix bug #394 'Copying a spy.StructDict returns a dict'.
```
Comment (tensionhead) [inline CHANGELOG.md:11]: Nach dem release ist vor dem release, indeed :grin:

File: syncopy/shared/tools.py [hunk 2]
```diff
@@ -48,6 +47,13 @@ def __str__(self):
             ppStr = "{}"
         return ppStr
 
+    def copy(self):
+        """Overwrite the .copy method of the parent 'dict' class, otherwise copy() will return a dict."""
+        obj = type(self).__new__(self.__class__)
```
Comment (tensionhead) [inline syncopy/shared/tools.py:52]: Ok that's as clean as it gets :+1:

File: syncopy/tests/test_tools.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    def test_structdict_from_dict(self):
+        my_dict = {'a': 0.5, 'b': 'test', 'c' : [1, 2, 3]}
+        cfg = spy.StructDict(my_dict)
+        assert type(cfg) == spy.shared.tools.StructDict
+        assert cfg.a == 0.5
+        assert cfg.b == "test"
+        assert cfg.c == [1, 2, 3]
+
+    def test_structdict_from_dict_copy(self):
```
Comment (tensionhead) [inline syncopy/tests/test_tools.py:43]: Is that test really necessary? Once the `cfg` is a `StructDict`, this test just repeats the standard copy test no?!

File: syncopy/tests/test_welch.py [hunk 4]
```diff
@@ -362,6 +349,10 @@ def test_welch_rejects_invalid_output(self):
                 with pytest.raises(SPYValueError, match="output"):
                     _ = spy.freqanalysis(cfg, self.adata)
 
+    def teardown_class(cls):
```
Comment (tensionhead) [inline syncopy/tests/test_welch.py:352]: yes, this we could probably need also elsewhere, maybe put it inside a `helper` module?

**Review by tensionhead — COMMENTED** [review]
Great, this will make the `StructDict` behavior more consistent.. just one question, how can we make deep copies? The default shallow copies seem quite dangerous from a user perspective?!

File: syncopy/tests/test_tools.py (hunk 3 above)
Comment (dfsp-spirit) [reply syncopy/tests/test_tools.py:43]: yes, I will remove it. I just had the impression in between that something was fishy, but it had a different reason.

Comment (dfsp-spirit) [reply syncopy/tests/test_tools.py]: its removed

File: syncopy/tests/test_welch.py (hunk 4 above)
Comment (dfsp-spirit) [reply syncopy/tests/test_welch.py:353]: I moved it to `tests/helpers.py`

**Comment by dfsp-spirit** [comment]
> 

I have implemented a `.deepcopy()` method in [9560672](https://github.com/esi-neuroscience/syncopy/commit/95606720c6ff381ffd72d9b31c0a8e9c676d56ce).

File: syncopy/shared/tools.py [hunk 5]
```diff
@@ -48,6 +48,42 @@ def __str__(self):
...
+
+        Notes
+        -----
+        Consider calling the `.deepcopy()` method to get a deep copy instead, which is typically what users want.
+        """
+        obj = type(self).__new__(self.__class__)
+        obj.__dict__.update(self.__dict__)
+        return obj
+
+    def deepcopy(self):
```
Comment (tensionhead) [inline syncopy/shared/tools.py:69]: Wondering if we should just wire `copy()` to this deepcopy method as a default?! I think from a user perspective shallow copies are hardly a useful feature, even though shallow copies are more or less Python's default?

**Review by tensionhead — APPROVED** [review]
Ok great, now ppl can safely throw around `.copy()` on their `cfg`'s 👍

**Pull request merged**